src/object/srv_ec_aggregate.c

```diff
@@ -0,0 +1,1203 @@
+	unsigned char		*parity_bufs[p];
+	unsigned char		*vects[3];
+	unsigned char		*buf = NULL;
+	unsigned char		*obuf = NULL;
+	unsigned char		*old = NULL;
+	unsigned char		*new = NULL;
+	unsigned char		*diff = NULL;
+	int			 i, j, rc = 0;
+
+	codec = obj_ec_codec_get(daos_obj_id2class(entry->ae_oid.id_pub));
+	buf = entry->ae_sgl->sg_iovs[AGG_IOV_PARITY].iov_buf;
+	for (i = 0; i < p; i++)
+		parity_bufs[i] = &buf[i*len];
+
+	obuf = entry->ae_sgl->sg_iovs[AGG_IOV_ODATA].iov_buf;
+	buf = entry->ae_sgl->sg_iovs[AGG_IOV_DATA].iov_buf;
+	diff = entry->ae_sgl->sg_iovs[AGG_IOV_DIFF].iov_buf;
+	for (i = 0, j = 0; i < cell_cnt; i++) {
+		old = &obuf[i*cell_bytes];
+		new = &buf[i*cell_bytes];
+		vects[0] = old;
+		vects[1] = new;
+		vects[2] = diff;
+		rc = xor_gen(3, cell_bytes, (void **)vects);
+		if (rc)
+			goto out;
+		while (!isset(bit_map, j))
+			j++;
+		ec_encode_data_update(cell_bytes, k, p, j, codec->ec_gftbls,
+				      diff, parity_bufs);
+	}
+out:
+	return rc;
+}
+
+/* Xstream offload function for partial stripe update. Fetches the old data
+ * from the data target(s) and updates the parity.
+ */
+static void
+agg_process_partial_stripe_ult(void *arg)
+{
+	struct ec_agg_stripe_ud		*stripe_ud =
+						(struct ec_agg_stripe_ud *)arg;
+	struct ec_agg_entry		*entry = stripe_ud->asu_agg_entry;
+	uint8_t				*bit_map = stripe_ud->asu_bit_map;
+	unsigned int			 cell_cnt = stripe_ud->asu_cell_cnt;
+	int				 rc = 0;
+
+	rc = agg_fetch_odata_cells(entry, bit_map, cell_cnt);
+	if (rc)
+		goto out;
+	rc = agg_update_parity(entry, bit_map, cell_cnt);
+
+out:
+	ABT_eventual_set(stripe_ud->asu_eventual, (void *)&rc, sizeof(rc));
+
+}
+
+/* Driver function for partial stripe update. Fetches the data and then invokes
+ * second function to update the parity.
+ */
+static int
+agg_process_partial_stripe(struct ec_agg_entry *entry)
+{
+	struct ec_agg_stripe_ud	 stripe_ud;
+	struct ec_agg_extent	*agg_extent;
+	unsigned int		len = entry->ae_oca->u.ec.e_len;
+	unsigned int		k = entry->ae_oca->u.ec.e_k;
+	unsigned int		i, cell_cnt = 0;
+	uint8_t			bit_map[OBJ_TGT_BITMAP_LEN] = {0};
+	int			*status;
+	int			rc = 0;
+
+	for (i = 0; i < k && cell_cnt < k; i++)
+		d_list_for_each_entry(agg_extent,
+				      &entry->ae_cur_stripe.as_dextents,
+				      ae_link) {
+			if (agg_overlap(&agg_extent->ae_recx, i, k, len,
+					entry->ae_cur_stripe.as_stripenum)) {
+				setbit(bit_map, i);
+				cell_cnt++;
+			}
+			if (cell_cnt == k)
+				break;
+		}
+
+
+	rc = agg_fetch_local_extents(entry, bit_map, cell_cnt);
+	if (rc)
+		goto out;
+
+	stripe_ud.asu_agg_entry = entry;
+	stripe_ud.asu_bit_map = bit_map;
+	stripe_ud.asu_cell_cnt = cell_cnt;
+	rc = ABT_eventual_create(sizeof(*status), &stripe_ud.asu_eventual);
+	if (rc != ABT_SUCCESS) {
+		rc = dss_abterr2der(rc);
+		goto out;
+	}
+	rc = dss_ult_create(agg_process_partial_stripe_ult, &stripe_ud,
+			    DSS_ULT_EC, 0, 0, NULL);
+	if (rc)
+		goto out;
+	rc = ABT_eventual_wait(stripe_ud.asu_eventual, (void **)&status);
```

Comment:
hmm, so you mean "moving to next stripe" has to wait current "local vos updating." why? I may miss sth here. And also this wait will block the whole container EC iteration?

Reply:
Right now it does, because I reuse the entry/stripe context structs when I move on to the next stripe.  I can redo the structs so the helper ULT can enqueue its completions.  Then the I/O ult could then process subsequent stripes, and also periodically flush the vos_update queue.

It's just the aggregation ULT that is blocked.  Other ULTs in the I/O Xstream (e.g., I/O RPC handlers) will be scheduled.

Reply:
yes, but if there is space pressure, it probably want to "reclaim" these space as soon as possible.  Hmm, ok, this probably can be optimized later. it